### ISSUE TITLE
Migration Error with Project Config

### ISSUE THREAD

**Issue by davist11** [doublesecretagency/craft-smartmap#40]
I am getting an error after updating to version 3.2.2 when running migrations on a staging server with project config enabled and admin changes disallowed:

```
Exception: Changes to the project config are not possible while in read-only mode. (/var/www/bdi/staging/releases/20190327161148/vendor/craftcms/cms/src/services/ProjectConfig.php:371)
#0 /var/www/bdi/staging/releases/20190327161148/vendor/craftcms/cms/src/services/Fields.php(765): craft\services\ProjectConfig->set('fields.5e8e07c3...', Array)
#1 /var/www/bdi/staging/releases/20190327161148/vendor/doublesecretagency/craft-smartmap/src/migrations/m190310_000000_smartMap_addPositionToFieldLayouts.php(118): craft\services\Fields->saveField(Object(doublesecretagency\smartmap\fields\Address), false)
#2 /var/www/bdi/staging/releases/20190327161148/vendor/craftcms/cms/src/db/Migration.php(56): doublesecretagency\smartmap\migrations\m190310_000000_smartMap_addPositionToFieldLayouts->safeUp()
#3 /var/www/bdi/staging/releases/20190327161148/vendor/craftcms/cms/src/db/MigrationManager.php(243): craft\db\Migration->up(true)
#4 /var/www/bdi/staging/releases/20190327161148/vendor/craftcms/cms/src/db/MigrationManager.php(163): craft\db\MigrationManager->migrateUp(Object(doublesecretagency\smartmap\migrations\m190310_000000_smartMap_addPositionToFieldLayouts))
#5 /var/www/bdi/staging/releases/20190327161148/vendor/craftcms/cms/src/services/Updates.php(225): craft\db\MigrationManager->up()
#6 /var/www/bdi/staging/releases/20190327161148/vendor/craftcms/cms/src/console/controllers/MigrateController.php(252): craft\services\Updates->runMigrations(Array)
#7 [internal function]: craft\console\controllers\MigrateController->actionAll()
#8 /var/www/bdi/staging/releases/20190327161148/vendor/yiisoft/yii2/base/InlineAction.php(57): call_user_func_array(Array, Array)
#9 /var/www/bdi/staging/releases/20190327161148/vendor/yiisoft/yii2/base/Controller.php(157): yii\base\InlineAction->runWithParams(Array)
#10 /var/www/bdi/staging/releases/20190327161148/vendor/yiisoft/yii2/console/Controller.php(148): yii\base\Controller->runAction('all', Array)
#11 /var/www/bdi/staging/releases/20190327161148/vendor/yiisoft/yii2/base/Module.php(528): yii\console\Controller->runAction('all', Array)
#12 /var/www/bdi/staging/releases/20190327161148/vendor/yiisoft/yii2/console/Application.php(180): yii\base\Module->runAction('migrate/all', Array)
#13 /var/www/bdi/staging/releases/20190327161148/vendor/craftcms/cms/src/console/Application.php(93): yii\console\Application->runAction('migrate/all', Array)
#14 /var/www/bdi/staging/releases/20190327161148/vendor/yiisoft/yii2/console/Application.php(147): craft\console\Application->runAction('migrate/all', Array)
#15 /var/www/bdi/staging/releases/20190327161148/vendor/yiisoft/yii2/base/Application.php(386): yii\console\Application->handleRequest(Object(craft\console\Request))
#16 /var/www/bdi/staging/releases/20190327161148/craft(22): yii\base\Application->run()
#17 {main}
Exception 'craft\errors\MigrateException' with message 'An error occurred while migrating Smart Map.'
```

Let me know if there are any other details I can provide.

**Comment by lindseydiloreto** [comment]
Hmm, hard to say how to solve this one. I don't believe the migration is actually doing anything too unusual. If the migration _needs_ to adjust the field settings, it inevitably would need to write to the project config. Seems like a bit of a catch-22.

I'll run this by the developer community, and maybe even ping P&T to get their advice.

Thanks for reporting it!

**Comment by lindseydiloreto** [comment]
Looks like we found the answer... the migration needs to cross-check your `project.yaml` folder before it decides whether or not to run.

<img width="769" alt="Ω 2019-03-28 at 11 54 54 AM" src="https://user-images.githubusercontent.com/5309692/55185113-e6079180-5150-11e9-82b7-a4658b7fd8f6.png">

Give me a day or two to implement it, I'll post back here with an update. 👍

**Comment by lindseydiloreto** [comment]
@davist11 Sorry for the delay on this.

Apparently, handling migrations for 3.1 is a whole different ballgame. I've been consulting with Andris to determine the best course of action moving forward... it may require that I bump up the plugin's minimum requirements to Craft 3.1 exclusively.

**What level of urgency do you have on this?** I want to get it fixed ASAP, but I'm getting on a plane tomorrow and will be gone for a week.

**Comment by davist11** [comment]
Not a huge rush on this, I rolled the version back, so it's not currently causing any issues.

**Comment by lindseydiloreto** [comment]
Ok great. If I don't get to it in the next 24 hours, I'll tackle it when I get back into town. 👍

**Comment by lindseydiloreto** [comment]
I've released 3.2.2.1, which neutralizes this migration. It's a pretty low-stakes migration, so it's fine if it doesn't get a chance to run right now. I'm going to take another stab at a similar migration soon, and try to fix the bugs that cropped up in this one. Apparently project config + migrations is a pretty complicated subject. Once I've properly addressed those issues, Smart Map will officially require a minimum of Craft 3.1 (instead of the current 3.0 minimum).

All that said, feel free to try updating again! I'm 99% confident that it will go smoothly. 🙂

**Comment by lindseydiloreto** [comment]
Alright, this has been officially fixed for the next release. 👍

Smart Map 3.2.3 (not yet released) will now require Craft 3.1 to be installed. It contains an updated version of this migration which smooths out this (and another unrelated) issue.
